HELP.md

```diff
@@ -0,0 +1,814 @@
+They all need to have unique prefixes.
+
+We can introduce an `enum` for tracking storage prefixes and keys.
+And then use borsh serialization to construct a unique prefix for every collection.
+It's as efficient as manually constructing them, because with Borsh serialization, an enum only takes one byte.
+
+```rust
+#[derive(BorshSerialize)]
+pub enum StorageKeys {
+    Accounts,
+    SubAccount { account_hash: Vec<u8> },
+    Tokens,
+    Metadata,
+}
+
+#[near_bindgen]
+impl Contract {
+    #[init]
+    pub fn new() -> Self {
+        Self {
+            accounts: UnorderedMap::new(StorageKeys::Accounts.try_to_vec().unwrap()),
+            tokens: LookupMap::new(StorageKeys::Tokens.try_to_vec().unwrap()),
+            metadata: LazyOption::new(StorageKeys::Metadata.try_to_vec().unwrap()),
+        }
+    }
+    
+    fn get_tokens(&self, account_id: &AccountId) -> UnorderedSet<String> {
+        let tokens = self.accounts.get(account_id).unwrap_or_else(|| {
+            UnorderedSet::new(
+                StorageKeys::SubAccount { account_hash: env::sha256(account_id.as_bytes()) }
+                    .try_to_vec()
+                    .unwrap(),
+            )
+        });
+        tokens
+    }
+}
+```
+
+For a traditional way of handling it, see [instructions below](#the-traditional-way-of-handling-unique-prefixes-for-persistent-collections)
+
+
+## Upgrading a contract
+
+After `3.0.1` change, `#[init]` macro initializes the contract and verifies that the old state doesn't exist.
+It will panic if the old state (under key `STATE`) is present in the storage.
+
+But if you need to re-initialize the contract STATE, you need to use `#[init(ignore_state)]` instead.
+This will NOT check that the state exists and you can use it in case you need to upgrade contract and migrate state.
+
+```rust
+#[near_bindgen]
+impl Contract {
+    #[init(ignore_state)]
+    pub fn migrate_state(new_data: String) -> Self {
+        // Deserialize the state using the old contract structure.
+        let old_contract: OldContract = env::state_read().expect("Old state doesn't exist");
+        // Verify that the migration can only be done by the owner.
+        // This is not necessary, if the upgrade is done internally.
+        assert_eq!(
+            &env::predecessor_account_id(),
+            &old_contract.owner_id,
+            "Can only be called by the owner"
+        );
+
+        // Create the new contract using the data from the old contract.
+        Self { owner_id: old_contract.owner_id, data: old_contract.data, new_data }
+    }
+}
+```
+
+## Use `PanicOnDefault`
+
+By default `near_sdk` allows a contract to be initialized with default state.
+Usually, if you have an initializer, you will want to prevent this.
+There is a helper derive macro `PanicOnDefault` to do this, e.g.
+
+```rust
+#[near_bindgen]
+#[derive(BorshDeserialize, BorshSerialize, PanicOnDefault)]
+pub struct Contract {
+    pub data: String,
+}
+```
```

Comment:
Why do we need a separate proc macro for this? seems like `near_bindgen` can generate `Default` impl  which panics if ther's a method annotated with `#[init]`? Or maybe, better still, the code-gen can be changed to not rely on `Default` at all if there's init? 

It feels a bit that this is a work-around rather a best practice :)